File: Stadium.java
package unl.cse;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class Stadium {

    String name;
    String teamName;
    String city;
    String state;
    int yearBuilt;
    double latitude;
    double longitude;

    public Stadium(String name,
                   String teamName,
                   String city,
                   String state,
                   int yearBuilt,
                   double latitude,
                   double longitude) {
        this.name = name;
        this.teamName = teamName;
        this.city = city;
        this.state = state;
        this.yearBuilt = yearBuilt;
        if(latitude > -180 && latitude < 180){
            this.latitude = latitude;
        }else{
            this.latitude = 0;
        }
        if(longitude > -180 && longitude < 180){
            this.longitude = longitude;
        }else{
            this.longitude = 0;
        }
    }

    public String getName(){
        return this.name;
    }

    public String getTeamName(){
        return this.teamName;
    }

    public String getCity(){
        return this.city;
    }

    public String getState(){
        return this.state;
    }

    public int getYearBuilt(){
        return this.yearBuilt;
    }

    public double getLatitude(){
        return this.latitude;
    }

    public double getLongitude(){
        return this.longitude;
    }

    @Override
    public String toString() {
        String outPut = name + ", home of the " + teamName + ", (" + yearBuilt
                + ") " + city + ", " + state + " (" + latitude + ", " + longitude + ")";
        return outPut;
    }

    public double getAirDistance(Stadium destination) {
        return Math.acos(Math.sin(Math.toRadians(this.latitude)) * Math.sin(Math.toRadians(destination.latitude))
                + Math.cos(Math.toRadians(this.latitude)) * Math.cos(Math.toRadians(destination.latitude))
                * Math.cos(Math.toRadians(destination.longitude) - Math.toRadians(this.longitude))) * 6371;
    }

    public static double getAirDistances(List<Stadium> stadiums) {
        double distance = 0;
        for(int i = 1; i < stadiums.size(); i++){
            distance += stadiums.get(i - 1).getAirDistance(stadiums.get(i));
        }
        return distance;
    }

    public static final Comparator<Stadium> byName = new Comparator<Stadium>() {
        @Override
        public int compare(Stadium arg0, Stadium arg1) {
            return arg0.getName().compareTo(arg1.getName());
        }
    };

    public static final Comparator<Stadium> byTeamName = new Comparator<Stadium>() {
        @Override
        public int compare(Stadium arg0, Stadium arg1) {
            return arg0.getTeamName().compareTo(arg1.getTeamName());
        }
    };

    public static final Comparator<Stadium> byAge = new Comparator<Stadium>() {
        @Override
        public int compare(Stadium arg0, Stadium arg1) {
            return Integer.compare(arg0.getYearBuilt(), arg1.getYearBuilt());
        }
    };

    public static final Comparator<Stadium> byAgeNewestFirst = new Comparator<Stadium>() {
        @Override
        public int compare(Stadium arg0, Stadium arg1) {
            return Integer.compare(arg1.getYearBuilt(), arg0.getYearBuilt());
        }
    };

    public static final Comparator<Stadium> byStateCity = new Comparator<Stadium>() {
        @Override
        public int compare(Stadium arg0, Stadium arg1) {
            int value = arg0.getState().compareTo(arg1.getState());
            if(value != 0){
                return value;
            }else{
                return arg0.getCity().compareTo(arg1.getCity());
            }
        }
    };

    public static final Comparator<Stadium> byLatitude = new Comparator<Stadium>() {
        @Override
        public int compare(Stadium arg0, Stadium arg1) {
            return Double.compare(arg0.getLatitude(), arg1.getLatitude());
        }
    };

    public static final Comparator<Stadium> byLongitude = new Comparator<Stadium>() {
        @Override
        public int compare(Stadium arg0, Stadium arg1) {
            return Double.compare(arg0.getLongitude(), arg1.getLongitude());
        }
    };

    public static void sortStadiums(List<Stadium> stadiums, Order order) {
        switch (order)
        {
            case NAME:
                Collections.sort(stadiums, byName);
                break;
            case TEAM:
                Collections.sort(stadiums, byTeamName);
                break;
            case AGE:
                Collections.sort(stadiums, byAge);
                break;
            case AGE_NEWEST:
                Collections.sort(stadiums, byAgeNewestFirst);
                break;
            case STATE_CITY:
                Collections.sort(stadiums, byStateCity);
                break;
            case LATITUDE:
                Collections.sort(stadiums, byLatitude);
                break;
            case LONGITUDE:
                Collections.sort(stadiums, byLongitude);
                break;
            default:
                System.out.println("Error, please input an order");
                return;
        }
    }

}

File: stadium.c
#include"stadium.h"
#include<string.h>
#include<math.h>
#include<stdlib.h>
#include<stdio.h>
#define M_PI 3.14159265358979323846
#define toDegrees(a) (a * 180 / M_PI)
#define toRadians(a) (a * M_PI / 180)
#define RAD_EARTH 6371

//builder function for a stadium struct
Stadium * createStadium(const char *name, const char *teamName, const char *city, const char *state, int yearBuilt, double latitude, double longitude) {
	Stadium *newStadium = (Stadium *)malloc(sizeof(Stadium));
	newStadium->name = (char *)malloc(sizeof(char) * (strlen(name) + 1));
	strcpy(newStadium->name, name);
	newStadium->teamName = (char *)malloc(sizeof(char) * (strlen(teamName) + 1));
	strcpy(newStadium->teamName, teamName);
	newStadium->city = (char *)malloc(sizeof(char) * (strlen(city) + 1));
	strcpy(newStadium->city, city);
	newStadium->state = (char *)malloc(sizeof(char) * (strlen(state) + 1));
	strcpy(newStadium->state, state);
	newStadium->yearBuilt = yearBuilt;
	if (latitude > -180 && latitude < 180) { //ensures latitude is valid
		newStadium->latitude = latitude;
	}
	else {
		newStadium->latitude = 0;
	}
	if (longitude > -180 && longitude < 180) { //ensures longitude is valid
		newStadium->longitude = longitude;
	}
	else {
		newStadium->longitude = 0;
	}
	return newStadium;
}

//this function returns a string that contains all the information specific to the passed in stadium
char * stadiumToString(const Stadium *s) {
	int size = snprintf(NULL, 0, "%s, home of the %s, (%d) %s, %s (%f, %f)", s->name, s->teamName, s->yearBuilt, s->city, s->state, s->latitude, s->longitude) + sizeof(char);
	char *outPut = (char *)malloc(size);
	size = snprintf(outPut, size, "%s, home of the %s, (%d) %s, %s (%f, %f)", s->name, s->teamName, s->yearBuilt, s->city, s->state, s->latitude, s->longitude);
	return outPut;
}

//using two stadium functions, this function returns the air distance between the two
double getAirDistance(const Stadium *s1, const Stadium *s2) {
	return acos(sin(toRadians(s1->latitude)) * sin(toRadians(s2->latitude)) + cos(toRadians(s1->latitude)) * cos(toRadians(s2->latitude)) * cos(toRadians(s2->longitude) - toRadians(s1->longitude))) * RAD_EARTH;
}

//when passed an array of stadium structs of size n, this function calculates and returns the air distance between every one.
double getAirDistances(const Stadium *stadiums, int n) {
	int i;
	double distance = 0;
	for (i = 0; i < n - 1; i++) {
		distance += getAirDistance(&(stadiums[i]), &(stadiums[i + 1])); //#reusecode
	}
	return distance;
}

//name comparator
int cmpStadiumByName(const void *s1, const void *s2) {
	Stadium *x = (Stadium *)s1;
	Stadium *y = (Stadium *)s2;
	return strcmp(x->name, y->name);
}

//team name comparator
int cmpStadiumByTeamName(const void*s1, const void*s2) {
	Stadium *x = (Stadium *)s1;
	Stadium *y = (Stadium *)s2;
	return strcmp(x->teamName, y->teamName);
}

//age comparator
int cmpStadiumByAge(const void*s1, const void*s2) {
	Stadium *x = (Stadium *)s1;
	Stadium *y = (Stadium *)s2;
	return ((x->yearBuilt - y->yearBuilt) == 0) ? 0 : ((x->yearBuilt - y->yearBuilt) / abs(x->yearBuilt - y->yearBuilt));
}

//reverse age comparator
int cmpStadiumByAgeNewestFirst(const void*s1, const void*s2) {
	return cmpStadiumByAge(s2, s1);
}

//state, then city comparator
int cmpStadiumByStateCity(const void*s1, const void*s2) {
	Stadium *x = (Stadium *)s1;
	Stadium *y = (Stadium *)s2;
	int returnValue = strcmp(x->state, y->state);
	return (!returnValue) ? strcmp(x->city, y->city) : returnValue;
}

//latitude comparator
int cmpStadiumByLatitude(const void*s1, const void*s2) {
	Stadium *x = (Stadium *)s1;
	Stadium *y = (Stadium *)s2;
	return ((x->latitude - y->latitude) == 0.0) ? 0 : (int)((x->latitude - y->latitude) / fabs(x->latitude - y->latitude));
}

//longitude comparator
int cmpStadiumByLongitude(const void*s1, const void*s2) {
	Stadium *x = (Stadium *)s1;
	Stadium *y = (Stadium *)s2;
	return ((x->longitude - y->longitude) == 0.0) ? 0 : (int)((x->longitude - y->longitude) / fabs(x->longitude - y->longitude));
}

//sorts the array of stadiums using the Order value and the above comparators 
void sortStadiums(Stadium *stadiums, int n, Order order) {
	int(*comparator)(const void*s1, const void*s2);
	switch (order) //sets comparator
	{
	case NAME:
		comparator = cmpStadiumByName;
		break;
	case TEAM:
		comparator = cmpStadiumByTeamName;
		break;
	case AGE:
		comparator = cmpStadiumByAge;
		break;
	case AGE_NEWEST:
		comparator = cmpStadiumByAgeNewestFirst;
		break;
	case STATE_CITY:
		comparator = cmpStadiumByStateCity;
		break;
	case LATITUDE:
		comparator = cmpStadiumByLatitude;
		break;
	case LONGITUDE:
		comparator = cmpStadiumByLongitude;
		break;
	default:
		printf("Error, please input an order\n");
		return;
	}
	qsort(stadiums, n, sizeof(Stadium), comparator);  //default sorting
}
